composer/datasets/in_context_learning_evaluation.py

```diff
@@ -0,0 +1,120 @@
+        for data_pair in data:
+            context, continuation = data_pair['context'], data_pair['continuation']
+
+            context_enc = context['input_ids']
+            continuation_enc = continuation['input_ids']
+            continuation_span = torch.tensor(range(len(context_enc), len(context_enc) + len(continuation_enc)))
+
+            inp = torch.tensor(
+                (context_enc + continuation_enc)[-(self.max_seq_len + 1):],
+                dtype=torch.long,
+            )
+            (inplen,) = inp.shape
+
+            # pad length from seq to padding_length
+            inp = torch.cat(
+                [
+                    inp,  # [seq]
+                    torch.LongTensor((self.max_seq_len - inplen) * [self.eos_tok_id]),  # [padding_length - seq]
+                ],
+                dim=0,
+            )
+
+            inputs.append(inp)
+            continuation_indices.append(continuation_span)
+
+        return {
+            'input_ids': torch.stack(inputs),
+            'continuation_indices': continuation_indices,
+            'mode': 'lm_task',
+            'labels': torch.stack(inputs),
+            'eos_tok_id': self.eos_tok_id
+        }
+
+    def get_num_samples_in_batch(self, batch) -> int:
+        if isinstance(batch, torch.Tensor):
+            return batch.shape[0]
+
+        dim0_sizes = []
+        if isinstance(batch, (list, tuple)):
+            for tensors in batch:
+                for t in ensure_tuple(tensors):
+                    if not hasattr(t, 'shape'):
+                        raise ValueError('Unable to determine the batch size, batch contains'
+                                         f'an element of type {type(t)}, which does not have a'
+                                         'shape. Please use a DataSpec and provide a'
+                                         '`get_num_samples_in_batch(your_batch) -> int` method.')
+                    dim0_sizes.append(t.shape[0])
+        elif isinstance(batch, dict):
+            dim0_sizes = [t.shape[0] for t in batch.values() if isinstance(t, torch.Tensor)]
+
+        if len(set(dim0_sizes)) == 1:
+            return dim0_sizes[0]
+        else:
+            raise NotImplementedError(
+                textwrap.dedent(f"""\
+                    Cannot determine the batch size, as multiple Tensors of
+                    different lengths were found in the batch: sizes in batch: {dim0_sizes}.
+                    Please use a DataSpec and specify `get_num_samples_in_batch`."""))
+
+    def update_metric(self, metric, batch, output_logits, labels):
+        metric.update(batch, output_logits, labels)
+
+
+def get_lm_task_dataloader(dataset_uri, tokenizer, batch_size, max_seq_len, eos_tok_id):
+    dataset = InContextLearningLMTaskDataset(dataset_uri, tokenizer, max_seq_len, eos_tok_id)
+    sampler = dist.get_sampler(dataset, drop_last=False, shuffle=True)
+    print(f'Using microbatch size {batch_size}')
+    return DataSpec(DataLoader(
```

Comment:
I think you don't need to wrap this in `DataSpec`. Did you find that you did need to for some reason?